lib/pf/Authentication/Source/PaypalSource.pm

```diff
@@ -0,0 +1,322 @@
+package pf::Authentication::Source::PaypalSource;
+
+=head1 NAME
+
+pf::Authentication::Source::PaypalSource
+
+=cut
+
+=head1 DESCRIPTION
+
+pf::Authentication::Source::PaypalSource
+
+=cut
+
+use strict;
+use warnings;
+use Moose;
+use pf::config qw($FALSE $TRUE $default_pid $fqdn);
+use pf::Authentication::constants;
+use pf::util;
+use pf::log;
+use WWW::Curl::Easy;
+use JSON::XS;
+use List::Util qw(first pairmap);
+use URI::Escape qw(uri_escape uri_unescape);
+use HTTP::Status qw(is_success :constants);
+use IPC::Open2;
+
+our $PAYPAL_SANDBOX_URL = 'https://www.sandbox.paypal.com/cgi-bin/webscr';
+our $PAYPAL_URL = 'https://www.paypal.com/cgi-bin/webscr';
+
+extends 'pf::Authentication::Source::BillingSource';
+
+=head2 Attributes
+
+=head2 class
+
+=cut
+
+has '+class' => (default => 'billing');
+
+has '+type' => (default => 'Paypal');
+
+has 'identity_token' => ( isa => 'Str', is => 'rw', required => 1);
+
+has 'domains' => (isa => 'Str', is => 'rw', required => 1, default => '*.paypal.com,*.paypalobjects.com');
+
+has 'cert_id' => ( isa => 'Str', is => 'rw', required => 1);
+
+has 'cert_file' => ( isa => 'Str', is => 'rw', required => 1);
+
+has 'key_file' => ( isa => 'Str', is => 'rw', required => 1);
+
+has 'paypal_cert_file' => ( isa => 'Str', is => 'rw', required => 1);
+
+has 'email_address' => (isa => 'Str', is => 'rw', required => 1);
+
+has 'payment_type' => (isa => 'Str', is => 'rw', required => 1);
+
+has 'url' => (is => 'rw', builder => '_get_paypal_url', lazy => 1);
+
+=head2 prepare_payment
+
+Prepare the payment from paypal
+
+=cut
+
+sub prepare_payment {
+    my ($self, $session, $tier, $params, $uri) = @_;
+    return {
+        encrypted => $self->encrypt_form($session, $tier, $params),
+    };
+}
+
+=head2 encrypt_form
+
+
+=cut
+
+sub encrypt_form {
+    my ($self, $session, $tier, $params) = @_;
+    my $cert_file = $self->cert_file;
+    my $key_file = $self->key_file;
+    my $paypal_cert_file = $self->paypal_cert_file;
+    my $cmd = "/usr/bin/openssl smime -sign -signer \"$cert_file\" "
```

Comment:
I'm not the original author of this one, so I will let @jrouzierinverse comment on this one